src/SignalR/common/Http.Connections/src/Internal/Transports/LongPollingServerTransport.cs

```diff
@@ -115,6 +117,21 @@ public async Task ProcessRequestAsync(HttpContext context, CancellationToken tok
         }
     }
 
+    private static void AddNoCacheHeaders(HttpResponse response)
+    {
+        if (response.Headers.CacheControl.Count == 0) {
+            response.Headers.CacheControl = "no-cache, no-store, must-revalidate";
+        }
+
+        if (response.Headers.Pragma.Count == 0) {
+            response.Headers.Pragma = "no-cache";
+        }
+
+        if (response.Headers.Expires.Count == 0) {
+            response.Headers.Expires = "0";
+        }
```

Comment:
Why are we checking if these headers have already been set? Also, would `response.Headers.CacheControl = "no-store"` be sufficient? Are there any CDNs that wouldn't respect that?

@Tratcher

Reply:
We do something similar in cookie auth. Client implementations aren't consistent so we set all possible headers:
https://github.com/dotnet/aspnetcore/blob/d84666f43f0f27381e607778d3308bf67715c507/src/Security/Authentication/Cookies/src/CookieAuthenticationHandler.cs#L409-L411

Note Expires = '0' isn't the right format, it should be a date.

Checking if they're set first would only matter if you thought they would be customized by another component.

Reply:
> Note Expires = '0' isn't the right format, it should be a date.

MDN says

> Invalid expiration dates with value 0 represent a date in the past and mean that the resource is already expired.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Expires

Reply:
Don't expect consistent client behavior when using invalid values.

Reply:
> Why are we checking if these headers have already been set?

For example, middleware may by configured as
```cs
app.Use((context, next) =>
{
    if (context.Request.Path == "/hub")
    {
        context.Response.Headers.CacheControl = "no-cache, max-age=0";
    }
    return next(context);
});
```
I  think, we shouldn't just replace one header value `no-cache, max-age=0` with another `no-cache, no-store`.

Reply:
I don't think we should bother checking the already-set response headers. Cookie auth doesn't do it, we don't do it for other headers, and it still doesn't provide a way to disable the headers entirely. If this is something that really matters, we can add an option not to set the cache control headers at all, but I don't think this will be necessary.

Reply:
>  it still doesn't provide a way to disable the headers entirely ... we can add an option not to set the cache control headers at all

Ah, got it.